app/addons/documents/mango/components/MangoQueryEditor.js

```diff
@@ -130,16 +131,54 @@ export default class MangoQueryEditor extends Component {
   notifyOnQueryError() {
     if (this.editorHasErrors()) {
       FauxtonAPI.addNotification({
-        msg:  'Please fix the Javascript errors and try again.',
+        msg:  'Please fix the JSON errors and try again.',
         type: 'error',
         clear: true
       });
 
       return true;
     }
+
     return false;
   }
 
+  getJsonIfValid(json) {
+    try {
+      return JSON.parse(json);
+    } catch (e) {
+      FauxtonAPI.addNotification({
+        msg:  'Please fix the JSON errors and try again. Error: ' + e.message,
+        type: 'error',
+        clear: true
+      });
+    }
+  }
+
+  isRegexValid(selector = {}) {
+    const regexes = Helper.getSelectorRegexes(selector);
+    const errors = _.reduce(regexes, (acc, val) => {
+      try {
+        new RegExp(val);
```

Comment:
it's worth bearing in mind that Query uses [Regex engine](http://erlang.org/doc/man/re.html) that is not perfectly compatible with JavaScript's. This check will allow through regex strings that will not work in Query and there may be regexes that are valid in Query but not valid JS regexes.

Is there a particular use case that led to this PR? For instance, do we need better server-side errors in the event that a bad regex is passed in a selector?

Reply:
See the linked issue. I guess the best solution would be to have more human friendly error message when having regex errors. If that's not possible, I could allow to send wrong regex to the server and use my code to return more detailed error message in case of failure

Reply:
@willholley @popojargo Right, the idea is that we need much better error messages from Couch when a regex fails, and that many ultra-simple syntax errors can be caught in client-side directly - for instance having sufficient `/` count.

I appreciate the "impedence mismatch" between Erlang re and JavaScript regexes being a concern. The ultimate goal is to enable people working in Fauxton to find their errors more readily, and for the editor provide some sort of syntax highlighting or in-line guide to what might be wrong vs. having to read a likely truncated, obscure error message in the warning bar at the top after submission.

Suggestions are very welcome here :) Personally, I'm reacting to requests for improvement from IRC/Slack support for basic things like regex syntax highlighting and error parsing, which seem like low hanging fruit, even if we take a naive approach. What do you think?

Reply:
Adding a custom highlighter for the $regex key-value pairs in the editor looks complicated to me. Although, even with a good highlighter, the error will still be vague.

So I guess the first thing could be to find a way to have more expressive error messages.

My two suggestions:

- Reformat the Regex error message (display the letters as char, not charCode).
- On a query error, we can check the server error and display an appropriate error message if it's caused by regex errors (I can reverse engineer the request and detect Regex errors with the JavaScript regex engine.